### PR TITLE
Using a topological sort to order kernels on OpenMP backend

### PR DESCRIPTION
**Pull request by TobyFlynn**
This pull request uses a topological sort to create a run list for the kernels on the OpenMP backend. This fixes an issue where dependencies between kernels might not be respected after setting a cache blocking group.

I've tested this on the AV test case I was using previously and on all cases in the PyFR test case repo with 10 MPI ranks on the OpenMP backend - they all look good. I've not touched any of the other backends but let me know if there are any further testing you'd like me to do for this pull request.

### CODE REVIEW

**Comment by FreddieWitherden** [comment]
Worth checking that all of the MPI requests start at a 'good' time (as in they are not all lumped towards the end) as this can hurt scaling.  Further, if we are doing a topological sort to determine the order then the OpemMP backend may need `pdeps`.

File: pyfr/backends/openmp/types.py [hunk 1]
```diff
@@ -55,6 +56,10 @@ class OpenMPXchgMatrix(OpenMPMatrix, base.XchgMatrix): pass
 class OpenMPXchgView(base.XchgView): pass
 class OpenMPView(base.View): pass
 
+# Need a hashable wrapper for a list of kernels for the topological sort
+class OpenMPGroupNode:
```
Comment (FreddieWitherden) [inline pyfr/backends/openmp/types.py:60]: See if we can avoid this by using `id`.

Comment (TobyFlynn) [reply pyfr/backends/openmp/types.py]: Is the latest commit more what you were thinking in terms of using `id`? If it is, I've only tested it quickly - I'll set off some more tests overnight and let you when they've finished before merging

**Comment by FreddieWitherden** [comment]
What would a solution not using pdeps or a topological sort look like?  So just trying to dynamically insert the group but while keeping as close as possible to the original ordering?

**Pull request closed without merging**